csharp/tools/Microsoft.ML.OnnxRuntime.PerfTool/Program.cs

```diff
@@ -47,106 +51,312 @@ public static void Main(string[] args)
-        {
-            var tensorData = new List<float>();
-
-            // read data from file
-            using (var inputFile = new System.IO.StreamReader(filename))
-            {
-                inputFile.ReadLine(); //skip the input name
-                string[] dataStr = inputFile.ReadLine().Split(new char[] { ',', '[', ']' }, StringSplitOptions.RemoveEmptyEntries);
-                for (int i = 0; i < dataStr.Length; i++)
-                {
-                    tensorData.Add(Single.Parse(dataStr[i]));
-                }
-            }
-
-            return tensorData.ToArray();
+            PrintReport(timestamps, timecosts, iteration);
+            Console.WriteLine("Done");
         }
 
-        static void RunModelOnnxRuntime(string modelPath, string inputPath, int iteration, DateTime[] timestamps, bool parallelExecution, GraphOptimizationLevel optLevel)
+        static void RunModelOnnxRuntime(string modelPath, string inputPath, int iteration, DateTime[] timestamps,
+                                        double[] timecosts, bool parallelExecution, GraphOptimizationLevel optLevel)
         {
             if (timestamps.Length != (int)TimingPoint.TotalCount)
             {
                 throw new ArgumentException("Timestamps array must have " + (int)TimingPoint.TotalCount + " size");
             }
 
+            Random random = new Random();
+
             timestamps[(int)TimingPoint.Start] = DateTime.Now;
             SessionOptions options = new SessionOptions();
             if (parallelExecution) options.ExecutionMode = ExecutionMode.ORT_PARALLEL;
             options.GraphOptimizationLevel = optLevel;
             using (var session = new InferenceSession(modelPath, options))
             {
                 timestamps[(int)TimingPoint.ModelLoaded] = DateTime.Now;
-                var inputMeta = session.InputMetadata;
 
-                var container = new List<NamedOnnxValue>();
-                foreach (var name in inputMeta.Keys)
+                var containers = LoadTestData(modelPath, inputPath, session.InputMetadata);
+                timestamps[(int)TimingPoint.InputLoaded] = DateTime.Now;
+
+                // Warm-up
                 {
-                    float[] rawData = LoadTensorFromFile(inputPath);
-                    var tensor = new DenseTensor<float>(rawData, inputMeta[name].Dimensions);
-                    container.Add(NamedOnnxValue.CreateFromTensor<float>(name, tensor));
+                    var container = containers[random.Next(0, containers.Count)];
+                    session.Run(container);
                 }
-
-
-
-                timestamps[(int)TimingPoint.InputLoaded] = DateTime.Now;
+                timestamps[(int)TimingPoint.WarmUp] = DateTime.Now;
 
                 // Run the inference
                 for (int i = 0; i < iteration; i++)
                 {
+                    var next = random.Next(0, containers.Count);
+                    var container = containers[next];
```

Comment:
nit: it would be great if user can choose the sample to run.

Reply:
User can organize test data directories like onnxruntime_perf_tool. When we have such a requirement, I'll add this feature later at both managed and native perf tool.